### PR TITLE
Allow specification of number of partitions in XShards.partition

### PR DESCRIPTION
**Pull request by GZHoffie**
I added a parameter `num_shards` that allows user to specify number of partitions in the `XShards.partition` function, and the issue #3466 can be resolved by changing the code to
```python
sample_shards = XShards.partition({"x": sampled}, num_shards=1)
```
so that there is no empty shards. Do I need to add a warning message if there is empty shard?

### CODE REVIEW

**Comment by yangw1234** [comment]
raise an error if `length` is smaller than the number of shards?

**Comment by yangw1234** [comment]
also need unit tests

**Comment by GZHoffie** [comment]
http://10.239.47.210:18888/job/ZOO-PR-Validation/5044/ Most builds shows that build is success, but with the following errors
```
PEP8 checks failed.
./zoo/examples/orca/learn/horovod/pytorch_estimator.py:120:101: E501 line too long (101 > 100 characters)
./zoo/examples/orca/learn/horovod/pytorch_estimator.py:131:101: E501 line too long (111 > 100 characters)
./zoo/examples/orca/learn/horovod/pytorch_estimator.py:133:101: E501 line too long (117 > 100 characters)
./zoo/examples/orca/learn/horovod/pytorch_estimator.py:135:101: E501 line too long (115 > 100 characters)
./zoo/examples/orca/learn/horovod/pytorch_estimator.py:137:101: E501 line too long (106 > 100 characters)
./zoo/examples/orca/learn/horovod/pytorch_estimator.py:139:19: E128 continuation line under-indented for visual indent
./zoo/examples/orca/learn/horovod/pytorch_estimator.py:140:19: E128 continuation line under-indented for visual indent
./zoo/examples/orca/learn/horovod/pytorch_estimator.py:141:19: E128 continuation line under-indented for visual indent
./zoo/examples/orca/learn/horovod/pytorch_estimator.py:142:19: E128 continuation line under-indented for visual indent
Build step 'Execute shell' marked build as failure
Finished: FAILURE
```
the tests failed. I guess this is not related to my PR and will check later for solutions.

**Comment by yangw1234** [comment]
you need to add one or more unit tests to cover this case.

**Review by yangw1234 — CHANGES_REQUESTED** [review]
unit tests needed

File: pyzoo/zoo/orca/data/shard.py [hunk 1]
```diff
@@ -61,16 +61,17 @@ def load_pickle(cls, path, minPartitions=None):
...
         Partition local in memory data and form a SparkXShards
         :param data: np.ndarray, a tuple, list, dict of np.ndarray, or a nested structure
         made of tuple, list, dict with ndarray as the leaf value
+        :param num_shards: the number of shards that the data will be partitioned into
         :return: a SparkXShards
         """
         sc = init_nncontext()
         node_num, core_num = get_node_and_core_number()
-        total_core_num = node_num * core_num
+        total_core_num = node_num * core_num if num_shards is None else num_shards
```
Comment (yangw1234) [inline pyzoo/zoo/orca/data/shard.py:74]: change `total_core_num` to something else, like "shard_num"

File: pyzoo/zoo/orca/data/shard.py [hunk 2]
```diff
@@ -82,13 +83,19 @@ def partition(data):
         """.format(type(data))
         supported_types = {list, tuple, dict}
         if isinstance(data, np.ndarray):
+            assert data_length >= total_core_num, "The length of data {} is smaller than " \
```
Comment (yangw1234) [inline pyzoo/zoo/orca/data/shard.py:86]: where this `data_length` is defined?

File: pyzoo/zoo/orca/data/shard.py [hunk 3]
```diff
@@ -82,13 +83,19 @@ def partition(data):
...
+                "the total number of shards {}. Please adjust the num_shards option to be " \
+                "at most {}.".format(data.shape[0], total_core_num, data.shape[0])
```
Comment (yangw1234) [inline pyzoo/zoo/orca/data/shard.py:88]: "at least"?

File: pyzoo/zoo/orca/data/shard.py (hunk 2 above)
Comment (GZHoffie) [reply pyzoo/zoo/orca/data/shard.py:86]: Sorry, it's a copy mistake... Should be `data.shape[0]`

File: pyzoo/zoo/orca/data/shard.py (hunk 3 above)
Comment (GZHoffie) [reply pyzoo/zoo/orca/data/shard.py:88]: Should be at most? If number of shards is too large then there will be empty shards.

Do we need to check empty shards here or in the tf_runner?

Comment (yangw1234) [reply pyzoo/zoo/orca/data/shard.py]: yeah, ok

**Review by yangw1234 — APPROVED** [review]

**Comment by GZHoffie** [comment]
Jenkins passed http://10.239.47.210:18888/job/ZOO-PR-Validation/5061/, the case `ZOO-PR-Validation-Spark-2.4` is run again and passed here http://10.239.47.210:18888/job/ZOO-PR-Validation-Spark-2.4/3419/

**Pull request merged**